# Introduction
In this notebook we are going to explore the information of the table "Spotify-Top-Songs" of the Database "Spotify-Songs" that we create for the project. We will also document the development of the project.

# Description of the project
For this project, we're going to make a Power BI dashboard. The dashboard will show the most popular songs on Spotify in each country. To develop this project, we're going to upload the dataset to a PostgreSQL database. We'll also use this notebook to preview the data and detect any issues, and then clean the table. After the data is cleaned, we will connect the database to Power BI to create the dashboard.

# Data Description
The data proportioned has the following columns of data:

    * spotify_id: Unique identifier for the song in Spotify
    * name: The title of the song
    * artist: Name(s) of the artist(s) associated with the song
    * daily_rank: Daily rank of the song in the top 50 list (0 to 50)
    * daily_movement: Change in rankings compared to the previous day (-49 to 50)
    * weekly_movement: Change in rankings compared to the previous week (-49 to 50)
    * country: ISO code of the country of the TOP 50 Playlist
    * snapshot_date: Date on which the data was collected from Spotify
    * popularity: Measure of the song's current popularity on Spotify (0 to 100)
    * is_explicit: Whether the song contains explicit lyrics
    * duration_ms: The duration of the song in milliseconds
    * album_name: Name of the album of the song
    * album_release_date: Date of the release of the album of the song
    * danceability
    * energy
    * key
    * loudness
    * mode
    * speechiness
    * acousticness
    * instrumentalness
    * liveness
    * valence
    * tempo
    * time_signature

# Procedure
We are going to segment this project in different sections:

    * Data Exploration

    * Data Cleaning

    * Dashboard Creation

Having clarified the above, let's get started.

# Data Exploration
In this section we are going to load the data to a database in PostgreSQL, we are going to connect to that database using Python and explore the data for errors and anomalies.

Data Load To PostgreSQL
When we load the data to our database, we assign this type of data to the columns:

    * spotify_id: Character(30)
    * name: Text
    * artist: Text
    * daily_rank: Integer
    * daily_movement: Integer
    * weekly_movement: Integer
    * country: Character(10)
    * snapshot_date: Date
    * popularity: Integer
    * is_explicit: Boolean
    * duration_ms: Integer
    * album_name: Text
    * album_release_date: Date
    * danceability: Double Precision
    * energy: Double Precision
    * key: Integer
    * loudness: Double Precision
    * mode: Integer
    * speechiness: Double Precision
    * acousticness: Double Precision
    * instrumentalness: Double Precision
    * liveness: Double Precision
    * valence: Double Precision
    * tempo: Double Precision
    * time_signature: Double Precision

But we couldn't assign the Date type to the column album_release_date, let's review why:

In [1]:
# Importing libraries
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# Connection to the data base
db_config = {'user': 'postgres',
                'pwd': '0123456789',
                'host': 'localhost',
                'port': '5432',
                'db': 'Spotify-Songs'}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], db_config['pwd'], db_config['host'], db_config['port'], db_config['db'])

engine = create_engine(connection_string)

In [3]:
# Obtaining the data of the table "spotify_top_songs"
query = """SELECT *
FROM "spotify_top_songs";
"""

df = pd.io.sql.read_sql(query, con = engine)

df.head(5)

spotify_id                    name  \
0  3tfiVsnpCzewE8rPpsRIL6                         عالعموم   
1  0rGCXktltsDt5jmcFwjFm4                     Aal Doghrey   
2  6XnBt8ME4K8d9oYf7SE5Wa           سوري مبردش انا بستوري   
3  0rDNeS6IwKi1oMrtkEc5S0                        Yazmeely   
4  4ACG3Z158G9E5NLfQ1CuHj          مهرجان لعبالي في دماغي   

                         artists  daily_rank  daily_movement  weekly_movement  \
0                         Shehab           6               0                0   
1                Reda El Bahrawy           7               0                0   
2  Magdy El Zahar, Eslam Kabonga           8               0                6   
3                          Muhab           9               1               -1   
4                    Moamen Moaa          10               3               32   

      country snapshot_date  popularity  is_explicit  ...  key loudness mode  \
0  EG            2024-09-09          65        False  ...   10  -11.283    0   
1  EG            2024-09-09          62        False  ...    7   -4.877    1   
2  EG            2024-09-09          58        False  ...    0   -0.513    1   
3  EG            2024-09-09          64        False  ...    5   -9.396    0   
4  EG            2024-09-09          50        False  ...    5    0.177    0   

   speechiness  acousticness  instrumentalness  liveness  valence    tempo  \
0       0.2530        0.2760          0.000000     0.123    0.406  116.996   
1       0.0316        0.1770          0.000002     0.036    0.914  122.020   
2       0.1730        0.4130          0.000000     0.261    0.947  110.078   
3       0.0401        0.5750          0.000000     0.112    0.224  100.045   
4       0.1420        0.0338          0.000000     0.329    0.870  111.116   

   time_signature  
0               4  
1               4  
2               4  
3               4  
4               4  

[5 rows x 25 columns]

In [ ]:
# Describing the data
display(df.info())
display(df.describe())
display(df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1728382 entries, 0 to 1728381
Data columns (total 25 columns):
 #   Column              Dtype  
---  ------              -----  
 0   spotify_id          object 
 1   name                object 
 2   artists             object 
 3   daily_rank          int64  
 4   daily_movement      int64  
 5   weekly_movement     int64  
 6   country             object 
 7   snapshot_date       object 
 8   popularity          int64  
 9   is_explicit         bool   
 10  duration_ms         int64  
 11  album_name          object 
 12  album_release_date  object 
 13  danceability        float64
 14  energy              float64
 15  key                 int64  
 16  loudness            float64
 17  mode                int64  
 18  speechiness         float64
 19  acousticness        float64
 20  instrumentalness    float64
 21  liveness            float64
 22  valence             float64
 23  tempo               float64
 24  time_signature      int6

None

daily_rank  daily_movement  weekly_movement    popularity  \
count  1.728382e+06    1.728382e+06     1.728382e+06  1.728382e+06   
mean   2.549208e+01    9.288537e-01     2.802143e+00  7.609990e+01   
std    1.442889e+01    7.011687e+00     1.215836e+01  1.574934e+01   
min    1.000000e+00   -4.900000e+01    -4.900000e+01  0.000000e+00   
25%    1.300000e+01   -1.000000e+00    -3.000000e+00  6.500000e+01   
50%    2.500000e+01    0.000000e+00     0.000000e+00  8.000000e+01   
75%    3.800000e+01    2.000000e+00     5.000000e+00  8.800000e+01   
max    5.000000e+01    4.900000e+01     4.900000e+01  1.000000e+02   

        duration_ms  danceability        energy           key      loudness  \
count  1.728382e+06  1.728382e+06  1.728382e+06  1.728382e+06  1.728382e+06   
mean   1.933514e+05  6.793745e-01  6.492625e-01  5.543001e+00 -6.622988e+00   
std    4.963848e+04  1.410755e-01  1.666627e-01  3.579551e+00  3.343594e+00   
min    0.000000e+00  0.000000e+00  2.010000e-05  0.000000e+00 -5.434100e+01   
25%    1.616550e+05  5.850000e-01  5.510000e-01  2.000000e+00 -7.805000e+00   
50%    1.859170e+05  7.010000e-01  6.680000e-01  6.000000e+00 -6.021000e+00   
75%    2.184230e+05  7.830000e-01  7.650000e-01  9.000000e+00 -4.710000e+00   
max    9.396660e+05  9.880000e-01  9.980000e-01  1.100000e+01  3.233000e+00   

               mode   speechiness  acousticness  instrumentalness  \
count  1.728382e+06  1.728382e+06  1.728382e+06      1.728382e+06   
mean   5.380558e-01  9.488872e-02  2.743085e-01      2.041868e-02   
std    4.985498e-01  9.088160e-02  2.509523e-01      1.043010e-01   
min    0.000000e+00  0.000000e+00  3.450000e-06      0.000000e+00   
25%    0.000000e+00  3.850000e-02  6.700000e-02      0.000000e+00   
50%    1.000000e+00  5.780000e-02  1.890000e-01      1.280000e-06   
75%    1.000000e+00  1.100000e-01  4.370000e-01      8.500000e-05   
max    1.000000e+00  9.390000e-01  9.960000e-01      9.950000e-01   

           liveness       valence         tempo  time_signature  
count  1.728382e+06  1.728382e+06  1.728382e+06    1.728382e+06  
mean   1.707527e-01  5.497221e-01  1.222080e+02    3.901194e+00  
std    1.248310e-01  2.295444e-01  2.807423e+01    4.045418e-01  
min    1.390000e-02  0.000000e+00  0.000000e+00    0.000000e+00  
25%    9.610000e-02  3.710000e-01  1.000160e+02    4.000000e+00  
50%    1.210000e-01  5.520000e-01  1.199600e+02    4.000000e+00  
75%    2.050000e-01  7.350000e-01  1.400820e+02    4.000000e+00  
max    9.780000e-01  9.920000e-01  2.360890e+02    5.000000e+00

spotify_id                             name  \
0  2ctjDCCg1wHoQSjIJ8p6U4                                    Candy   
1  0DWdj2oZMBFSzRsi2Cvfzf                                      TQG   
2  7JbMsR4rZh6J77LNafur8U                     ¿por Que Te Demoras?   
3  69Ej1xrGjOcHvIMtMKxK0G                                     Dile   
4  06qMRF18gwbOYYbnP2du6i          Last Christmas - Single Version   

            artists  daily_rank  daily_movement  weekly_movement     country  \
0            Plan B          47              -5               -1  SV           
1  KAROL G, Shakira          48               1              -10  SV           
2            Plan B          49               1                1  SV           
3          Don Omar          50              -5               -5  SV           
4             Wham!           1               0                0  SK           

  snapshot_date  popularity  is_explicit  ...  key loudness mode  speechiness  \
0    2023-12-16          75        False  ...   11   -3.735    0       0.1470   
1    2023-12-16          89         True  ...    4   -3.547    0       0.2770   
2    2023-12-16          63        False  ...    2   -5.923    0       0.0769   
3    2023-12-16          84        False  ...    4   -7.501    0       0.1410   
4    2023-12-16          88        False  ...    2   -8.228    1       0.0278   

   acousticness  instrumentalness  liveness  valence    tempo  time_signature  
0        0.2520          0.000000    0.5170    0.542   95.981               4  
1        0.6730          0.000000    0.0936    0.607  179.974               4  
2        0.0386          0.000000    0.0547    0.941   96.018               4  
3        0.1840          0.000132    0.0420    0.714   94.001               4  
4        0.2120          0.000004    0.1560    0.935  107.732               4  

[5 rows x 25 columns]

At first instance, we don't see any anomaly but let's review if the all the rows have the same date format.

In [9]:
# Corroborating if the column "album_release_date" have the same format for all it's rows
query = """SELECT album_release_date
FROM "spotify_top_songs"
WHERE album_release_date !~ '^\d{4}-\d{2}-\d{2}$';
"""

pd.io.sql.read_sql(query, con = engine)

album_release_date
0                     
1                     
2                     
3                     
4                     
..                 ...
653                   
654                   
655                   
656                   
657                   

[658 rows x 1 columns]

As we can se, there are 658 rows that doesn't have information, but in the inicial description of the data we didn't find any missing data, the explanation for this is that the rows it isn't empty at all, it contains "". Let's view which songs have this issue.

In [14]:
# Reviewing the songs that have this "album_release_date" issue:
query = """SELECT name, artists, album_name, COUNT(album_release_date)
FROM "spotify_top_songs"
WHERE album_release_date !~ '^\d{4}-\d{2}-\d{2}$'
GROUP BY name, artists, album_name
ORDER BY name DESC;
"""

pd.io.sql.read_sql(query, con = engine)

name        artists  \
0                       Silent Night - 1999 Remaster  Frank Sinatra   
1              Mistletoe And Holly - Remastered 1999  Frank Sinatra   
2                     Jingle Bells - Remastered 1999  Frank Sinatra   
3  Have Yourself A Merry Little Christmas - Remas...  Frank Sinatra   
4                                    Brown Eyed Girl   Van Morrison   
5                                                                     

  album_name  count  
0                 3  
1                 1  
2               200  
3                92  
4               333  
5                29

As we can see, there are there are only five songs that have this issue and there is a song that not even have a name, we can get rid of it. But first of all, let's make that the empty rows in the data shows like "Null" elements so that way we can detect them more easily. So let's identify the columns that have empty rows with Python:

In [4]:
# identifying columns that have empty rows
for column in df.columns:
    if (df[column] == "").sum() > 0:
        print("The column {} has {} empty rows".format(column, (df[column] == "").sum()))

As we can see, there are only four columns that have this problem, let's fix that problem for that we will run this line in our database: 

```sql
-- Replacing "" for NULL
UPDATE spotify_top_songs
SET 
    "name" = CASE WHEN "name" = '' THEN NULL ELSE "name" END,
    artists = CASE WHEN artists = '' THEN NULL ELSE artists END,
    album_name = CASE WHEN album_name = '' THEN NULL ELSE album_name END,
    album_release_date = CASE WHEN album_release_date = '' THEN NULL ELSE album_release_date END;
```

This query, runs through the columns `name`, `artists`, `album_name` and `album_release_date` and looks for those empty rows and replace them with NULL.

Once we did this, let's review the data once more in python:

In [4]:
# Obtaining the data of the table "spotify_top_songs"
query = """SELECT *
FROM "spotify_top_songs";
"""

df = pd.io.sql.read_sql(query, con = engine)

In [8]:
# Describing the data:
display(df.isna().sum())

spotify_id              0
name                   30
artists                29
daily_rank              0
daily_movement          0
weekly_movement         0
country                 0
snapshot_date           0
popularity              0
is_explicit             0
duration_ms             0
album_name            821
album_release_date    658
danceability            0
energy                  0
key                     0
loudness                0
mode                    0
speechiness             0
acousticness            0
instrumentalness        0
liveness                0
valence                 0
tempo                   0
time_signature          0
dtype: int64

Now the data is ready to begin the cleaning process.

# Data Cleaning

In the previous section we find that there are some empty rows in the columns `name`, `artists`, `album_name`, `album_name` and `album_release_date`. Let's review the rows which the columns `name` and `artist` are NULL:

In [9]:
# Selecting the rows which `name` and `artists` are NULL
query = """
SELECT
    * 
FROM 
    spotify_top_songs 
WHERE
    "name" IS NULL OR
    artists IS NULL;
"""

pd.io.sql.read_sql(query, con = engine)

spotify_id  name artists  daily_rank  daily_movement  \
0   7lyv2sysHCzFjypILxAynT          None    None          36              14   
1   7lyv2sysHCzFjypILxAynT          None    None          18              32   
2   7lyv2sysHCzFjypILxAynT          None    None           5              45   
3   7lyv2sysHCzFjypILxAynT          None    None          32              18   
4   7lyv2sysHCzFjypILxAynT          None    None          46               4   
5   6yxtsR3nc3aUL1wcbLn8A3          None    None          30               1   
6   35wSewUpyeXwde8wVlHW7L          None    NOVR           3              47   
7   3vz3SKnCyJCzPuXWLoGfdG          None    None           6              44   
8   2h4b8QdmU4nxZrlpz7INIs          None    None           1              49   
9   7lyv2sysHCzFjypILxAynT          None    None          21              29   
10  0kvD9ksvXyRHANPypIpkIh          None    None          48               2   
11  0kvD9ksvXyRHANPypIpkIh          None    None          25              25   
12  4nLJ61BP9owXw9sLoLtqIa          None    None           2              48   
13  7lyv2sysHCzFjypILxAynT          None    None          29              21   
14  7lyv2sysHCzFjypILxAynT          None    None          12              38   
15  0kvD9ksvXyRHANPypIpkIh          None    None          20              30   
16  3GMeCx87zWbtg6jD4evZ58          None    None          22              28   
17  0wbcG4kKNOdZ9wTEra3aYI          None    None          45               5   
18  3vz3SKnCyJCzPuXWLoGfdG          None    None          20              30   
19  2V2K1hzCtgj9xAnga9WUTy          None    None          18              -1   
20  7lyv2sysHCzFjypILxAynT          None    None          25              25   
21  7lyv2sysHCzFjypILxAynT          None    None          17              33   
22  7lyv2sysHCzFjypILxAynT          None    None          42               8   
23  7lyv2sysHCzFjypILxAynT          None    None           9              41   
24  7lyv2sysHCzFjypILxAynT          None    None          11              39   
25  0kvD9ksvXyRHANPypIpkIh          None    None          43               7   
26  0kvD9ksvXyRHANPypIpkIh          None    None          48               2   
27  0kvD9ksvXyRHANPypIpkIh          None    None          37              13   
28  0kvD9ksvXyRHANPypIpkIh          None    None          37              13   
29  0y2sUHVDAWQbfuIQGEfhFO          None    None           2              48   

    weekly_movement     country snapshot_date  popularity  is_explicit  ...  \
0                14  VE            2023-11-10           0         True  ...   
1                32  SV            2023-11-10           0         True  ...   
2                45  PE            2023-11-10           0         True  ...   
3                18  PA            2023-11-10           0         True  ...   
4                 4  NI            2023-11-10           0         True  ...   
5                20  NG            2023-10-21           0        False  ...   
6                47  EE            2024-10-28           0        False  ...   
7                44  AR            2024-03-14           0        False  ...   
8                49  IS            2024-01-18           0        False  ...   
9                29  CR            2023-11-10           0         True  ...   
10                2  SV            2023-11-09           0         True  ...   
11               25  PE            2023-11-09           0         True  ...   
12               48  MA            2025-01-15           0        False  ...   
13               21  GT            2023-11-10           0         True  ...   
14               38  ES            2023-11-10           0         True  ...   
15               30  ES            2023-11-09           0         True  ...   
16               28  TR            2024-04-27           0         True  ...   
17                5  TR            2024-04-27           0         True  ...   
18               30  UY            2024-03

As we can see, the rows which the column "name" and "artists" are Null are almost the same and also the columns "album_name" and "album_release_date" are NULL too.

This 30 rows only represents the 0.0017% of all our data, so we can get rid of them. So we execute this query in our data base: 

```sql
DELETE FROM spotify_top_songs
WHERE
    "name" IS NULL OR
    artists IS NULL;
```

Once we did this procedure, let's review the others empty rows, those are in the columns `album_name` and `album_release_date`.

In [8]:
# Reviewing the empty rows in the columns "album_name" and "album_release_date":
query = """
SELECT name, artists, album_name, album_release_date 
FROM spotify_top_songs
WHERE 
    album_name IS NULL OR
    album_release_date IS NULL;
"""

pd.io.sql.read_sql(query, con = engine)

name       artists album_name album_release_date
0    Brown Eyed Girl  Van Morrison       None               None
1    Brown Eyed Girl  Van Morrison       None               None
2               ABCD        NAYEON       None         2024-06-14
3    Brown Eyed Girl  Van Morrison       None               None
4               ABCD        NAYEON       None         2024-06-14
..               ...           ...        ...                ...
786  Brown Eyed Girl  Van Morrison       None               None
787  Brown Eyed Girl  Van Morrison       None               None
788  Brown Eyed Girl  Van Morrison       None               None
789  Brown Eyed Girl  Van Morrison       None               None
790             ABCD        NAYEON       None         2024-06-14

[791 rows x 4 columns]

In [9]:
# Grouping by song name
query = """
SELECT 
    "name",
    artists,
    SUM(CASE WHEN "album_name" IS NULL THEN 1 ELSE 0 END) AS "album_name_nulls",
    SUM(CASE WHEN "album_release_date" IS NULL THEN 1 ELSE 0 END) AS "album_release_date_nulls"
FROM 
    spotify_top_songs
WHERE
    album_name IS NULL OR
    album_release_date IS NULL
GROUP BY
    "name", artists
ORDER BY "name" DESC;
"""

pd.io.sql.read_sql(query, con = engine)

name        artists  \
0                       Silent Night - 1999 Remaster  Frank Sinatra   
1              Mistletoe And Holly - Remastered 1999  Frank Sinatra   
2                     Jingle Bells - Remastered 1999  Frank Sinatra   
3  Have Yourself A Merry Little Christmas - Remas...  Frank Sinatra   
4                                    Brown Eyed Girl   Van Morrison   
5                                               ABCD         NAYEON   

   album_name_nulls  album_release_date_nulls  
0                 3                         3  
1                 1                         1  
2               200                       200  
3                92                        92  
4               333                       333  
5               162                         0

As we can see there are only five songs that have missing a release date of the album and six songs that doesn't have an album name. Searching for the information we find that the songs:

* Silent Night - 1999 Remaster
* Mistletoe And Holly - Remastered 1999
* Jingle Bells - Remastered 1999
* Have Yourself A Merry Little Christmas - Remastered

Belongs to the same album "A Jolly Christmas from Frank Sinatra" and is released on "September 21 of 1957".

For the song: 

* Brown Eyed Girl

It belong to the album "Blowin' Your Mind!" and is released on "September 01 of 1967".

And lastly for the song: 

* ABCD

It belongs to the album "Na".

Having this information we can update our database with the following query: 

```sql
UPDATE  spotify_top_songs
SET
    album_name = CASE 
        WHEN artists = 'Frank Sinatra' AND album_name IS NULL THEN 'A Jolly Christmas from Frank Sinatra' 
        WHEN artists = 'Van Morrison' AND album_name IS NULL THEN 'Blowin'' Your Mind!'
        WHEN artists = 'NAYEON' AND album_name IS NULL THEN 'Na' 
        ELSE album_name
    END,
    album_release_date = CASE 
        WHEN artists = 'Frank Sinatra' AND album_release_date IS NULL THEN '1957-09-21'
        WHEN artists = 'Van Morrison' AND album_release_date IS NULL THEN '1967-09-01' 
        ELSE album_release_date 
    END;
```

Now that we already modify the data, let's corroborate the changes.

In [12]:
query = """
SELECT 
    "name",
    artists,
    album_name,
    album_release_date,
    COUNT(album_name),
    COUNT(album_release_date)
FROM
    spotify_top_songs
WHERE
    album_name LIKE 'A Jolly Christmas from Frank Sinatra' OR
    album_name LIKE 'Blowin'' Your Mind!' OR
    album_name LIKE 'Na'
GROUP BY
    "name", artists, album_name, album_release_date
ORDER BY
    "name" DESC;
"""

pd.io.sql.read_sql(query, con = engine)

name        artists  \
0                       Silent Night - 1999 Remaster  Frank Sinatra   
1              Mistletoe And Holly - Remastered 1999  Frank Sinatra   
2                     Jingle Bells - Remastered 1999  Frank Sinatra   
3  Have Yourself A Merry Little Christmas - Remas...  Frank Sinatra   
4                                    Brown Eyed Girl   Van Morrison   
5                                               ABCD         NAYEON   

                             album_name album_release_date  count  count  
0  A Jolly Christmas from Frank Sinatra         1957-09-21      3      3  
1  A Jolly Christmas from Frank Sinatra         1957-09-21      1      1  
2  A Jolly Christmas from Frank Sinatra         1957-09-21    200    200  
3  A Jolly Christmas from Frank Sinatra         1957-09-21     92     92  
4                    Blowin' Your Mind!         1967-09-01    333    333  
5                                    Na         2024-06-14    162    162

Finally, we didn't notice that the column `country` also have empty rows, but why it didn't show up in when we search for empty rows? let's review the unique values of that column.

In [ ]:
# Reviewing unique values of the column "country"
df.country.unique()

array(['EG        ', 'CO        ', 'AU        ', 'EE        ',
       'VE        ', 'EC        ', 'DO        ', 'DK        ',
       'DE        ', 'CZ        ', 'CR        ', 'CL        ',
       'CH        ', 'CA        ', 'BY        ', 'BR        ',
       'BO        ', 'BG        ', 'PA        ', 'BE        ',
       'AT        ', 'AR        ', 'AE        ', '          ',
       'ZA        ', 'VN        ', 'HN        ', 'UY        ',
       'US        ', 'UA        ', 'TW        ', 'TR        ',
       'TH        ', 'SE        ', 'SV        ', 'LV        ',
       'SK        ', 'SG        ', 'SA        ', 'RO        ',
       'PY        ', 'PT        ', 'PL        ', 'PK        ',
       'PH        ', 'PE        ', 'NZ        ', 'NO        ',
       'NL        ', 'NI        ', 'KZ        ', 'NG        ',
       'MY        ', 'MX        ', 'MA        ', 'LU        ',
       'JP        ', 'LT        ', 'KR        ', 'IT        ',
       'IS        ', 'IN        ', 'IL        ', 'IE   

As we can see, we have a unique value that doesn't contain any information, so we are going to rename the empty value for "Unknown", so:

```sql
UPDATE spotify_top_songs
SET
    country = CASE WHEN country = '' THEN 'Unknown' ELSE country END;
```

Now that we have our data complemented and cleaned, we can continue with the next step.

# Dashboard Creation

Now that we have our data clean, let's create the dashboard.

For our dashboard we are going to show the top 10 songs for a specific country (of course you could filter the country), the dashboard will show the popularity of those top 10 and the daily movement that the songs have, leaving us with the following result:

![Dashboard](./Images/dashboard_sc.png)

With this dashboard in place, we can now say that the project is complete.